src/lib/iptables.ts

```diff
@@ -1,66 +1,170 @@
-import * as Bluebird from 'bluebird';
-import * as childProcess from 'child_process';
-
-// The following is exported so that we stub it in the tests
-export const execAsync: (args: string) => Bluebird<string> = Bluebird.promisify(
-	childProcess.exec,
-);
-
-function applyIptablesArgs(args: string): Bluebird<void> {
-	let err: Error | null = null;
-	// We want to run both commands regardless, but also rethrow an error
-	// if one of them fails
-	return execAsync(`iptables ${args}`)
-		.catch((e) => (err = e))
-		.then(() => execAsync(`ip6tables ${args}`).catch((e) => (err = e)))
-		.then(() => {
-			if (err != null) {
-				throw err;
-			}
-		});
+import * as _ from 'lodash';
+import { child_process } from 'mz';
+
+export class IPTablesRuleError extends Error {}
+
+export interface Rule {
+	id?: number;
+	family?: 4 | 6;
+	action?: '-L' | '-I' | '-A' | '-D' | '-C' | '-N' | '-F' | '-X';
+	target?: 'ACCEPT' | 'BLOCK' | 'REJECT' | string;
+	chain?: string;
+	table?: 'filter' | string;
+	proto?: 'all' | any;
+	src?: string;
+	dest?: string;
+	matches?: string[];
+	applyIfExists?: boolean;
+	comment?: string;
 }
 
-function clearIptablesRule(rule: string): Bluebird<void> {
-	return applyIptablesArgs(`-D ${rule}`);
+export interface RuleBuilder {
+	add: (rules: Rule | Rule[]) => RuleBuilder;
+	apply: (adaptor: RuleAdaptor) => Promise<void>;
 }
 
-function clearAndAppendIptablesRule(rule: string): Bluebird<void> {
-	return clearIptablesRule(rule)
-		.catchReturn(null)
-		.then(() => applyIptablesArgs(`-A ${rule}`));
+export type RuleAdaptor = (rule: Rule) => Promise<string>;
+export function getDefaultRuleAdaptor(): RuleAdaptor {
+	return iptables;
 }
 
-function clearAndInsertIptablesRule(rule: string): Bluebird<void> {
-	return clearIptablesRule(rule)
-		.catchReturn(null)
-		.then(() => applyIptablesArgs(`-I ${rule}`));
+const iptables: RuleAdaptor = async (rule: Rule) => {
+	const args: string[] = [];
+
+	if (rule.action) {
+		args.push(rule.action);
+	}
+	if (rule.chain) {
+		args.push(rule.chain);
+	}
+	if (rule.action === '-L') {
+		args.push('--line-numbers');
+		args.push('-n');
+	}
+	if (rule.table) {
+		args.push(`-t ${rule.table}`);
+	}
+	if (rule.proto) {
+		args.push(`-p ${rule.proto}`);
+	}
+	if (rule.matches) {
+		rule.matches.forEach((match) => args.push(match));
+	}
+	if (rule.comment) {
+		args.push('-m comment');
+		args.push(`--comment "${rule.comment}"`);
+	}
+	if (rule.target) {
+		args.push(`-j ${rule.target}`);
+	}
+
+	if (!rule.family) {
+		rule.family = 4;
```

Comment:
Weren't we going to apply to both by default?

Reply:
We do, just not here -- this is the adaptor which applies the rule and it _needs_ a family is one is missing. However, we don't call this externally, we have another function which wraps/sanitises the *Rule* object and it handles the extra apply if it's missing a family 👍